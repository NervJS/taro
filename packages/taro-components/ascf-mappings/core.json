{
  "Button": {
    "size": {
      "ascfProp": "dimension",
      "valueMap": {
        "mini": "XS",
        "small": "S",
        "default": "M",
        "large": "L"
      },
      "required": false,
      "default": "M",
      "docLink": "https://ascf.docs/components/button#dimension"
    },
    "loading": {
      "ascfProp": "isLoading",
      "type": "boolean",
      "required": false,
      "default": false,
      "docLink": "https://ascf.docs/components/button#isloading"
    }
  },
  "Input": {
    "password": {
      "ascfProp": "secretInput",
      "type": "boolean",
      "required": false,
      "default": false,
      "docLink": "https://ascf.docs/components/input#secretinput"
    }
  }
}
