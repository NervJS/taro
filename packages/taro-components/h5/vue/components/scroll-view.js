import { listeners } from '../mixins/listeners'
import { refs } from '../mixins/refs'
export default {
  name: 'taro-scroll-view',
  mixins: [listeners, refs],
  props: {
    scrollX: Boolean,
    scrollY: Boolean
  },
  render (createElement) {
    // eslint-disable-next-line @typescript-eslint/no-this-alias
    const self = this

    const attrs = {}
    if (self.scrollX) attrs['scroll-x'] = true
    if (self.scrollY) attrs['scroll-y'] = true

    return createElement('taro-scroll-view-core', {
      class: ['hydrated', {
        'taro-scroll-view__scroll-x': self.scrollX,
        'taro-scroll-view__scroll-y': self.scrollY
      }],
      attrs,
      on: {
        ...self.listeners,
        scroll (e) {
          if (e instanceof CustomEvent) {
            self.$emit('scroll', e)
          }
        }
      }
    }, self.$slots.default)
  }
}
