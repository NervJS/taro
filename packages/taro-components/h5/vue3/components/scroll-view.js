import { h } from 'vue'

export default {
  emits: ['tap', 'scroll'],
  setup (props, { slots, emit, attrs }) {
    return () => (
      h(
        'taro-scroll-view-core',
        {
          class: ['hydrated', {
            'taro-scroll-view__scroll-x': attrs['scroll-x'] === '',
            'taro-scroll-view__scroll-y': attrs['scroll-y'] === ''
          }],
          onClick (e) {
            emit('tap', e)
          },
          onScroll (e) {
            if (e instanceof CustomEvent) {
              emit('scroll', e)
            }
          }
        },
        slots
      )
    )
  }
}
