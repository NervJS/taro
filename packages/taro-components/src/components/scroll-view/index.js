import 'weui'
import Nerv from 'nervjs'
import omit from 'omit.js'
import classNames from 'classnames'
import './style/index.scss'

function easeOutScroll (from = 0, to = 0, callback) {
  if (from === to || typeof from !== 'number') {
    return
  }
  let change = to - from
  const dur = 500
  const sTime = +new Date()
  function linear (t, b, c, d) {
    return c * t / d + b
  }
  const isLarger = to >= from

  function step () {
    from = linear(+new Date() - sTime, from, change, dur)
    if ((isLarger && from >= to) || (!isLarger && to >= from)) {
      callback(to)
      return
    }
    callback(from)
    requestAnimationFrame(step)
  }
  step()
}
function throttle (fn, delay) {
  let timer = null
  return function (...arrs) {
    clearTimeout(timer)
    timer = setTimeout(function () {
      fn(...arrs)
    }, delay)
  }
}
function scrollIntoView (id) {
  document.querySelector(`#${id}`).scrollIntoView({
    behavior: 'smooth',
    block: 'center',
    inline: 'start'
  })
}
function scrollVertical (top, isAnimation) {
  if (isAnimation) {
    easeOutScroll(this._scrollTop, top, pos => {
      if (this.container) this.container.scrollTop = pos
    })
  } else {
    if (this.container) this.container.scrollTop = top
  }
  this._scrollTop = top
}
function scrollHorizontal (left, isAnimation) {
  if (isAnimation) {
    easeOutScroll(this._scrollLeft, left, pos => {
      if (this.container) this.container.scrollLeft = pos
    })
  } else {
    if (this.container) this.container.scrollLeft = left
  }
  this._scrollLeft = left
}
class ScrollView extends Nerv.Component {
  constructor () {
    super(...arguments)
  }

  onTouchMove = e => {
    e.stopPropagation()
  }

  componentDidMount () {
    this.handleScroll(this.props, true)
  }

  componentWillReceiveProps (nextProps) {
    this.handleScroll(nextProps)
  }

  handleScroll (props, isInit = false) {
    // scrollIntoView
    if (
      props.scrollIntoView &&
      typeof props.scrollIntoView === 'string' &&
      document &&
      document.querySelector &&
      document.querySelector(`#${props.scrollIntoView}`)
    ) {
      if (isInit) {
        setTimeout(() => scrollIntoView(props.scrollIntoView), 500)
      } else {
        scrollIntoView(props.scrollIntoView)
      }
    } else {
      const isAnimation = 'scrollWithAnimation' in props
      // Y 轴滚动
      if (
        props.scrollY &&
        typeof props.scrollTop === 'number' &&
        props.scrollTop !== this._scrollTop
      ) {
        if (isInit) {
          setTimeout(() => scrollVertical.bind(this)(props.scrollTop, isAnimation), 10)
        } else {
          scrollVertical.bind(this)(props.scrollTop, isAnimation)
        }
      }
      // X 轴滚动
      if (
        props.scrollX &&
        typeof props.scrollLeft === 'number' &&
        props.scrollLeft !== this._scrollLeft
      ) {
        if (isInit) {
          setTimeout(() => scrollHorizontal.bind(this)(props.scrollLeft, isAnimation), 10)
        } else {
          scrollHorizontal.bind(this)(props.scrollLeft, isAnimation)
        }
      }
    }
  }

  render () {
    const {
      className,
      onScroll,
      onScrollToUpper,
      onScrollToLower,
      onTouchMove,
      scrollX,
      scrollY
    } = this.props
    let { upperThreshold = 50, lowerThreshold = 50 } = this.props
    const cls = classNames(
      'taro-scroll',
      {
        [`taro-scroll-view__scroll-x`]: scrollX,
        [`taro-scroll-view__scroll-y`]: scrollY
      },
      className
    )
    upperThreshold = parseInt(upperThreshold)
    lowerThreshold = parseInt(lowerThreshold)
    const uperAndLower = (e) => {
      if (!this.container) return
      const {
        offsetWidth,
        offsetHeight,
        scrollLeft,
        scrollTop,
        scrollHeight,
        scrollWidth
      } = this.container
      if (
        onScrollToLower &&
        ((this.props.scrollY &&
          offsetHeight + scrollTop + lowerThreshold >= scrollHeight) ||
          (this.props.scrollX &&
            offsetWidth + scrollLeft + lowerThreshold >= scrollWidth))
      ) {
        onScrollToLower(e)
      }
      if (
        onScrollToUpper &&
        ((this.props.scrollY && scrollTop <= upperThreshold) ||
          (this.props.scrollX && scrollLeft <= upperThreshold))
      ) {
        onScrollToUpper(e)
      }
    }
    const uperAndLowerThrottle = throttle(uperAndLower, 200)
    const _onScroll = e => {
      const {
        scrollLeft,
        scrollTop,
        scrollHeight,
        scrollWidth
      } = this.container
      this._scrollLeft = scrollLeft
      this._scrollTop = scrollTop
      Object.defineProperty(e, 'detail', {
        enumerable: true,
        writable: true,
        value: {
          scrollLeft,
          scrollTop,
          scrollHeight,
          scrollWidth
        }
      })
      uperAndLowerThrottle(e)
      onScroll && onScroll(e)
    }
    const _onTouchMove = e => {
      onTouchMove ? onTouchMove(e) : this.onTouchMove(e)
    }
    return (
      <div
        ref={container => {
          this.container = container
        }}
        {
        ...omit(this.props, ['className', 'scrollTop', 'scrollLeft'])
        }
        className={cls}
        onScroll={_onScroll}
        onTouchMove={
          _onTouchMove
        }
        onLoad={e => {
          console.log('onload', e)
        }} >
        {this.props.children}
      </div>
    )
  }
}

export default ScrollView
