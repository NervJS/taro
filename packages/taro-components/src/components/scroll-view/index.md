scroll-view

## API

|             | 属性                  | 类型        | 默认值 | 说明                                                                                 |
| ----------- | --------------------- | ----------- | ------ | ------------------------------------------------------------------------------------ |
| √           | scroll-x              | Boolean     | false  | 允许横向滚动                                                                         |
| √           | scroll-y              | Boolean     | false  | 允许纵向滚动                                                                         |
| √           | upper-threshold       | Number      | 50     | 距顶部/左边多远时（单位 px），触发 scrolltoupper 事件                                |
| √           | lower-threshold       | Number      | 50     | 距底部/右边多远时（单位 px），触发 scrolltolower 事件                                |
| √           | scroll-top            | Number      |        | 设置竖向滚动条位置                                                                   |
| √           | scroll-left           | Number      |        | 设置横向滚动条位置                                                                   |
|             | scroll-into-view      | String      |        | 值应为某子元素 id（id 不能以数字开头）。设置哪个方向可滚动，则在哪个方向滚动到该元素 |
| √           | scroll-with-animation | Boolean     | false  | 在设置滚动条位置时使用动画过渡                                                       |
|             | enable-back-to-top    | Boolean     | false  | iOS 点击顶部状态栏、安卓双击标题栏时，滚动条返回顶部，只支持竖向                     |
| √           | bindscrolltoupper     | EventHandle |        | 滚动到顶部/左边，会触发 scrolltoupper 事件                                           |
| √           | bindscrolltolower     | EventHandle |        | 滚动到底部/右边，会触发 scrolltolower 事件                                           |
| √(onScroll) | bindscroll            | EventHandle |        | 是否解码                                                                             |
