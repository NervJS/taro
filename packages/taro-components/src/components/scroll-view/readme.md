# taro-scroll-view-core

## API

|             | 属性                  | 类型        | 默认值 | 说明                                                                                 |
| ----------- | --------------------- | ----------- | ------ | ------------------------------------------------------------------------------------ |
| √           | scroll-x              | Boolean     | false  | 允许横向滚动                                                                         |
| √           | scroll-y              | Boolean     | false  | 允许纵向滚动                                                                         |
| √           | upper-threshold       | Number      | 50     | 距顶部/左边多远时（单位 px），触发 scrolltoupper 事件                                |
| √           | lower-threshold       | Number      | 50     | 距底部/右边多远时（单位 px），触发 scrolltolower 事件                                |
| √           | scroll-top            | Number      |        | 设置竖向滚动条位置                                                                   |
| √           | scroll-left           | Number      |        | 设置横向滚动条位置                                                                   |
|             | scroll-into-view      | String      |        | 值应为某子元素 id（id 不能以数字开头）。设置哪个方向可滚动，则在哪个方向滚动到该元素 |
| √           | scroll-with-animation | Boolean     | false  | 在设置滚动条位置时使用动画过渡                                                       |
|             | enable-back-to-top    | Boolean     | false  | iOS 点击顶部状态栏、安卓双击标题栏时，滚动条返回顶部，只支持竖向                     |
| √           | bindscrolltoupper     | EventHandle |        | 滚动到顶部/左边，会触发 scrolltoupper 事件                                           |
| √           | bindscrolltolower     | EventHandle |        | 滚动到底部/右边，会触发 scrolltolower 事件                                           |
| √(onScroll) | bindscroll            | EventHandle |        | 是否解码                                                                             |

<!-- Auto Generated Below -->


## Properties

| Property                    | Attribute                    | Description | Type                                        | Default     |
| --------------------------- | ---------------------------- | ----------- | ------------------------------------------- | ----------- |
| `animated`                  | `scroll-with-animation`      |             | `boolean`                                   | `false`     |
| `lowerThreshold`            | `lower-threshold`            |             | `number \| string`                          | `50`        |
| `mpScrollIntoView`          | `scroll-into-view`           |             | `string`                                    | `undefined` |
| `mpScrollIntoViewAlignment` | `scroll-into-view-alignment` |             | `"center" \| "end" \| "nearest" \| "start"` | `undefined` |
| `mpScrollLeft`              | `scroll-left`                |             | `number \| string`                          | `undefined` |
| `mpScrollTop`               | `scroll-top`                 |             | `number \| string`                          | `undefined` |
| `scrollX`                   | `scroll-x`                   |             | `boolean`                                   | `false`     |
| `scrollY`                   | `scroll-y`                   |             | `boolean`                                   | `false`     |
| `upperThreshold`            | `upper-threshold`            |             | `number \| string`                          | `50`        |


## Events

| Event           | Description | Type               |
| --------------- | ----------- | ------------------ |
| `scroll`        |             | `CustomEvent<any>` |
| `scrolltolower` |             | `CustomEvent<any>` |
| `scrolltoupper` |             | `CustomEvent<any>` |


## Methods

### `mpScrollIntoViewMethod(selector: string) => Promise<void>`



#### Returns

Type: `Promise<void>`



### `mpScrollToMethod(object: ScrollViewContext.scrollTo.Option) => Promise<void>`



#### Returns

Type: `Promise<void>`




----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
