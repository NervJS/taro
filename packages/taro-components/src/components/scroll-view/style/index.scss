.taro-scroll {
  -webkit-overflow-scrolling: touch;
  &::-webkit-scrollbar {
    display: none;
  }
}
.taro-scroll-view {
  overflow: hidden;
  &__scroll-x {
    overflow-x: scroll;
    overflow-y: hidden;
  }
  &__scroll-y {
    overflow-x: hidden;
    overflow-y: scroll;
  }
}
