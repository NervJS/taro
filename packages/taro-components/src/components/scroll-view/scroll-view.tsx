// eslint-disable-next-line @typescript-eslint/no-unused-vars
import { Component, h, ComponentInterface, Prop, Event, EventEmitter, Watch, Element, Host } from '@stencil/core'
import classNames from 'classnames'

function easeOutScroll (from: number, to: number, callback) {
  if (from === to || typeof from !== 'number') {
    return
  }

  const change = to - from
  const dur = 500
  const sTime = Date.now()
  const isLarger = to >= from

  function linear (t: number, b: number, c: number, d: number): number {
    return c * t / d + b
  }

  function step () {
    from = linear(Date.now() - sTime, from, change, dur)
    if ((isLarger && from >= to) || (!isLarger && to >= from)) {
      callback(to)
      return
    }
    callback(from)
    requestAnimationFrame(step)
  }

  step()
}

function debounce (fn, delay: number) {
  let timer: NodeJS.Timeout

  return function (...arrs) {
    clearTimeout(timer)
    timer = setTimeout(function () {
      fn(...arrs)
    }, delay)
  }
}

@Component({
  tag: 'taro-scroll-view-core',
  styleUrl: './style/index.scss'
})
export class ScrollView implements ComponentInterface {
  private _scrollLeft: number
  private _scrollTop: number

  @Element() el: HTMLElement

  @Prop() scrollX = false
  @Prop() scrollY = false
  @Prop() upperThreshold: number | string = 50
  @Prop() lowerThreshold: number | string = 50
  @Prop({ attribute: 'scroll-top' }) mpScrollTop: number | string
  @Prop({ attribute: 'scroll-left' }) mpScrollLeft: number | string
  @Prop({ attribute: 'scroll-into-view' }) mpScrollIntoView: string
  @Prop() scrollWithAnimation = false

  @Event({
    eventName: 'scroll'
  })
  onScroll: EventEmitter

  @Event({
    eventName: 'scrolltoupper'
  })
  onScrollToUpper: EventEmitter

  @Event({
    eventName: 'scrolltolower'
  })
  onScrollToLower: EventEmitter

  @Watch('mpScrollLeft')
  watchScrollLeft (newVal) {
    const scrollLeft = Number(newVal)
    if (
      this.scrollX &&
      !isNaN(scrollLeft) &&
      scrollLeft !== this._scrollLeft
    ) {
      if (this.scrollWithAnimation) {
        easeOutScroll(this._scrollLeft, scrollLeft, pos => (this.el.scrollLeft = pos))
      } else {
        this.el.scrollLeft = scrollLeft
      }
      this._scrollLeft = scrollLeft
    }
  }

  @Watch('mpScrollTop')
  watchScrollTop (newVal) {
    const scrollTop = Number(newVal)
    if (
      this.scrollY &&
      !isNaN(scrollTop) &&
      scrollTop !== this._scrollTop
    ) {
      if (this.scrollWithAnimation) {
        easeOutScroll(this._scrollTop, scrollTop, pos => (this.el.scrollTop = pos))
      } else {
        this.el.scrollTop = scrollTop
      }
      this._scrollTop = scrollTop
    }
  }

  @Watch('mpScrollIntoView')
  watchScrollIntoView (newVal) {
    if (typeof newVal === 'string' && newVal) {
      document.querySelector(`#${newVal}`)?.scrollIntoView({
        behavior: 'smooth',
        block: 'center',
        inline: 'start'
      })
    }
  }

  componentDidLoad () {
    const {
      scrollY,
      scrollX,
      scrollWithAnimation
    } = this

    const scrollTop = Number(this.mpScrollTop)
    const scrollLeft = Number(this.mpScrollLeft)

    if (scrollY && !isNaN(scrollTop)) {
      if (scrollWithAnimation) {
        easeOutScroll(0, scrollTop, pos => (this.el.scrollTop = pos))
      } else {
        this.el.scrollTop = scrollTop
      }
      this._scrollTop = scrollTop
    }

    if (scrollX && !isNaN(scrollLeft)) {
      if (scrollWithAnimation) {
        easeOutScroll(0, scrollLeft, pos => (this.el.scrollLeft = pos))
      } else {
        this.el.scrollLeft = scrollLeft
      }
      this._scrollLeft = scrollLeft
    }
  }

  handleScroll = (e) => {
    if (e instanceof CustomEvent) return

    const {
      scrollLeft,
      scrollTop,
      scrollHeight,
      scrollWidth
    } = this.el
    this._scrollLeft = scrollLeft
    this._scrollTop = scrollTop

    this.uperAndLower()

    this.onScroll.emit({
      scrollLeft,
      scrollTop,
      scrollHeight,
      scrollWidth
    })
  }

  uperAndLower = debounce(() => {
    const {
      offsetWidth,
      offsetHeight,
      scrollLeft,
      scrollTop,
      scrollHeight,
      scrollWidth
    } = this.el

    const lowerThreshold = Number(this.lowerThreshold)
    const upperThreshold = Number(this.upperThreshold)

    if (
      !isNaN(lowerThreshold) &&
      ((this.scrollY && offsetHeight + scrollTop + lowerThreshold >= scrollHeight) ||
      (this.scrollX && offsetWidth + scrollLeft + lowerThreshold >= scrollWidth))
    ) {
      this.onScrollToLower.emit({
        direction: this.scrollX ? 'right' : (this.scrollY ? 'bottom' : '')
      })
    }

    if (
      !isNaN(upperThreshold) &&
      ((this.scrollY && scrollTop <= upperThreshold) ||
      (this.scrollX && scrollLeft <= upperThreshold))
    ) {
      this.onScrollToUpper.emit({
        direction: this.scrollX ? 'left' : (this.scrollY ? 'top' : '')
      })
    }
  }, 200)

  render () {
    const {
      scrollX,
      scrollY
    } = this

    const cls = classNames({
      'taro-scroll-view__scroll-x': scrollX,
      'taro-scroll-view__scroll-y': scrollY
    })
    return (
      <Host
        class={cls}
        onScroll={this.handleScroll}
      >
        <slot />
      </Host>
    )
  }
}
