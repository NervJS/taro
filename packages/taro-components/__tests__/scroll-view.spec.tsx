import { h } from '@stencil/core'
import { newSpecPage, SpecPage } from '@stencil/core/testing'

import { ScrollView } from '../src/components/scroll-view/scroll-view'

describe('ScrollView', () => {
  let page: SpecPage
  // beforeAll(() => {
  //   // eslint-disable-next-line no-undef
  //   originTimeout = jasmine.DEFAULT_TIMEOUT_INTERVAL
  //   // eslint-disable-next-line no-undef
  //   jasmine.DEFAULT_TIMEOUT_INTERVAL = 40000
  // })

  // afterAll(() => {
  //   // eslint-disable-next-line no-undef
  //   jasmine.DEFAULT_TIMEOUT_INTERVAL = originTimeout
  // })

  const divStyleH = {
    display: 'inline-block',
    width: '100%',
    height: '300px'
  }

  const divStyleV = {
    height: '300px'
  }

  it('scroll-x', async () => {
    // TODO
    const onScroll = jest.fn()
    const onScrollToUpper = jest.fn()
    const onScrollToLower = jest.fn()
    page = await newSpecPage({
      components: [ScrollView],
      template: () => (<taro-scroll-view-core
        scrollX
        style={{
          'white-space': 'nowrap'
        }}
        onScroll={onScroll}
        onScrollToUpper={onScrollToUpper}
        onScrollToLower={onScrollToLower}
      >
        <div style={Object.assign({}, divStyleH, { backgroundColor: 'blue' })}></div>
        <div style={Object.assign({}, divStyleH, { backgroundColor: 'yellow' })}></div>
        <div style={Object.assign({}, divStyleH, { backgroundColor: 'blue' })}></div>
      </taro-scroll-view-core>),
    })
    // const { scrollHeight, scrollWidth } = node
    // const { width } = node.getBoundingClientRect()
    // const upper = 50
    // const lower = 50

    // expect(node.classList.contains('taro-scroll-view__scroll-x') === true)
    // expect(node.upperThreshold === upper)
    // expect(node.lowerThreshold === lower)
    // expect(node.scrollWithAnimation !== true)

    // await wrapper.setProps({
    //   scrollLeft: upper + 1,
    //   scrollWithAnimation: true
    // })

    // expect(onScroll.calledOnceWith({
    //   scrollLeft: upper + 1,
    //   scrollTop: 0,
    //   scrollHeight,
    //   scrollWidth
    // }))
    // expect(onScrollToUpper.callCount === 0)
    // expect(node.scrollWithAnimation === true)

    // await wrapper.setProps({
    //   scrollLeft: upper
    // })

    // expect(onScrollToUpper.calledOnceWith({ direction: 'left' }))

    // await wrapper.setProps({
    //   scrollLeft: scrollWidth - width - lower - 1
    // })

    // expect(onScrollToLower.callCount === 0)

    // await wrapper.setProps({
    //   scrollLeft: scrollWidth - width - lower
    // })

    // expect(onScrollToLower.calledOnceWith({ direction: 'right' }))

    expect(page.root).toMatchSnapshot()
  })

  it('scroll-y', async () => {
    // TODO
    const upper = 100
    const lower = 150
    const onScroll = jest.fn()
    const onScrollToUpper = jest.fn()
    const onScrollToLower = jest.fn()
    page = await newSpecPage({
      components: [ScrollView],
      template: () => (<taro-scroll-view-core
        scrollY
        style={{
          height: '300px'
        }}
        upperThreshold={upper}
        lowerThreshold={lower}
        onScroll={onScroll}
        onScrollToUpper={onScrollToUpper}
        onScrollToLower={onScrollToLower}
      >
        <div style={Object.assign({}, divStyleV, { backgroundColor: 'blue' })}></div>
        <div style={Object.assign({}, divStyleV, { backgroundColor: 'yellow' })}></div>
        <div style={Object.assign({}, divStyleV, { backgroundColor: 'blue' })}></div>
      </taro-scroll-view-core>),
    })
    // const { scrollHeight, scrollWidth } = node
    // const { height } = node.getBoundingClientRect()

    // expect(node.classList.contains('taro-scroll-view__scroll-y') === true)
    // expect(node.upperThreshold === upper)
    // expect(node.lowerThreshold === lower)

    // await wrapper.setProps({
    //   scrollTop: upper + 1
    // })

    // expect(onScroll.calledOnceWith({
    //   scrollLeft: 0,
    //   scrollTop: upper + 1,
    //   scrollHeight,
    //   scrollWidth
    // }))
    // expect(onScrollToUpper.callCount === 0)

    // await wrapper.setProps({
    //   scrollTop: upper
    // })

    // expect(onScrollToUpper.calledOnceWith({ direction: 'top' }))

    // await wrapper.setProps({
    //   scrollTop: scrollHeight - height - lower - 1
    // })

    // expect(onScrollToLower.callCount === 0)

    // await wrapper.setProps({
    //   scrollTop: scrollHeight - height - lower
    // })

    // expect(onScrollToLower.calledOnceWith({ direction: 'bottom' }))

    expect(page.root).toMatchSnapshot()
  })
})
