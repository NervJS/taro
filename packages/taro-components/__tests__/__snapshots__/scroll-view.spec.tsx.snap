// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScrollView scroll-x 1`] = `
<taro-scroll-view-core class="taro-scroll-view__scroll-x" style="white-space: nowrap;">
  <!---->
  <div style="display: inline-block; width: 100%; height: 300px; background-color: blue;"></div>
  <div style="display: inline-block; width: 100%; height: 300px; background-color: yellow;"></div>
  <div style="display: inline-block; width: 100%; height: 300px; background-color: blue;"></div>
</taro-scroll-view-core>
`;

exports[`ScrollView scroll-y 1`] = `
<taro-scroll-view-core class="taro-scroll-view__scroll-y" style="height: 300px;">
  <!---->
  <div style="height: 300px; background-color: blue;"></div>
  <div style="height: 300px; background-color: yellow;"></div>
  <div style="height: 300px; background-color: blue;"></div>
</taro-scroll-view-core>
`;
